dex doc (#491)
solo-io/solo-docs: gloo/docs/gloo_routing/virtual_services/authentication/external_auth/oauth/dex/_index.md

---
title: Dex and Gloo
weight: 2
description: Integrating Gloo and Dex Identity Provider
---

## Motivation

> [Dex Identify Provider](https://github.com/dexidp/dex) is an OpenID Connect identity hub. Dex can be used to expose a consistent OpenID Connect interface to your applications while allowing your users to use their existing identity from various back-ends, include LDAP, SAML, and other OIDC providers.

Reviewer: nit, wording:
```suggestion
[Dex Identify Provider](https://github.com/dexidp/dex) is an OpenID Connect identity hub. Dex can be used to expose a consistent OpenID Connect interface to your applications while allowing your users to use their existing identity provider from various back-ends, including LDAP, SAML, and other OIDC providers.
```

Using an Identity Hub like Dex has a few advantages:

- JWT based authentication lends itself well for distributed systems. With an identity hub you can re-use your existing investment

> - Allows you to change your authentication back-end without effecting the rest of the system.

Reviewer: nit, grammar:
```suggestion
- Allows you to change your authentication back-end without affecting the rest of the system.
```

> You can also use Dex for kubernetes itself. for example, to allow LDAP logins to work with kubectl.  

Reviewer: nit: grammar or capitalization
```suggestion
You can also use Dex for Kubernetes itself; for example, to allow LDAP logins to work with kubectl.  
```

This is outside the scope of this document, but you can read more about it [here](https://github.com/dexidp/dex/blob/master/Documentation/kubernetes.md).

In this document we will demonstrate how to integrate Gloo and Dex using Gloo's support for OpenID Connect.
This will allow using Dex to authenticate end users of Gloo's VirtualServices.

> For simplicity, this document will focus on deployment with a local cluster (like [minikube](https://github.com/kubernetes/minikube), or [kind](https://github.com/kubernetes-sigs/kind)) . With small changes these can be applied to a real cluster. We will use Dex with self sign certificate, as they are auto-generated. The same flow

Reviewer: nit grammar:
```suggestion
For simplicity, this document will focus on deployment with a local cluster (like [minikube](https://github.com/kubernetes/minikube), or [kind](https://github.com/kubernetes-sigs/kind)) . With small changes these can be applied to a real cluster. We will use Dex with self-signed certificates, as they are auto-generated. The same flow
```

> will work with user provided the certificates.

Reviewer: nit grammar:
```suggestion
will work with user-provided certificates.
```

## Install Dex
We will install Dex into the `gloo-system` namespace, and setup the alt-name in the Dex certificate to the 
correct service DNS name (so that later Gloo will trust the Dex service).
```
cat > /tmp/dex-values.yaml <<EOF

https: true
config:
  issuer: https://dex.gloo-system.svc.cluster.local:32000

  staticClients:
  - id: gloo
    redirectURIs:
    - 'http://localhost:8080/callback'
    name: 'GlooApp'
    secret: secretvalue
  
  staticPasswords:
  - email: "admin@example.com"
    # bcrypt hash of the string "password"
    hash: "\$2a\$10\$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
    username: "admin"
    userID: "08a8684b-db88-4b73-90a9-3cd1661f5466"

certs:
  web:
    altNames:
    - https://dex.gloo-system.svc.cluster.local:32000
    - dex.gloo-system.svc.cluster.local

EOF

helm install --name dex --namespace gloo-system stable/dex -f /tmp/dex-values.yaml
```

## Setup Dex CA in Gloo
Let's setup Gloo's external auth container to trust the Dex Certificate Authority.
We will add an init container that adds the Dex CA cert to the trusted CA certificates.

{{% notice note %}}
You may need to modify the command below to match your version of Gloo.

> You can edit your deployment and copy the highlighted parts.

Reviewer: we could also just provide a `kubectl edit` command to copy/paste here

        name: extauth
        env:
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: GLOO_ADDRESS
            value: gloo:9977
          - name: SIGNING_KEY
            valueFrom:
              secretKeyRef:
                name: extauth-signing-key
                key: signing-key
          - name: SERVER_PORT
            value: "8083"
          - name: USER_ID_HEADER
            value: "x-user-id"
          - name: START_STATS_SERVER
            value: "true"
      imagePullSecrets:
        - name: solo-io-readerbot-pull-secret
      affinity:
        podAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            - weight: 100
              podAffinityTerm:
                labelSelector:
                  matchLabels:
                    gloo: gateway-proxy
                topologyKey: kubernetes.io/hostname
EOF
{{< /highlight >}}

## Test!

### Deploy Demo App
Deploy the pet clinic demo app

```shell
kubectl --namespace default apply -f https://raw.githubusercontent.com/solo-io/gloo/v0.8.4/example/petclinic/petclinic.yaml
```

### Create a Virtual Service

> Create Gloo VirtualService with OIDC authenticatino enabled. Please note that the OIDC configuration below matches the one defined in Dex's staticClients config stanza:

Reviewer: nit: spelling
```suggestion
Create Gloo VirtualService with OIDC authentication enabled. Please note that the OIDC configuration below matches the one defined in Dex's staticClients config stanza:
```

>   the `oidc-auth-callback-path` mataches the `redirectURIs`, the `oidc-auth-client-id` matches the `id` and the `oidc-auth-client-secret` matches the `secret`.

Reviewer: nit: spelling
```suggestion
  the `oidc-auth-callback-path` matches the `redirectURIs`, the `oidc-auth-client-id` matches the `id` and the `oidc-auth-client-secret` matches the `secret`.
```

```
glooctl create  secret oauth --client-secret secretvalue oauth
glooctl create virtualservice --oidc-auth-app-url http://localhost:8080/ --oidc-auth-callback-path /callback --oidc-auth-client-id gloo --oidc-auth-client-secret-name oauth --oidc-auth-client-secret-namespace gloo-system --oidc-auth-issuer-url https://dex.gloo-system.svc.cluster.local:32000/ oidc-test --namespace gloo-system --enable-oidc-auth
```

Add a route to the pet clinic demo app.
```
glooctl add route --name default --namespace gloo-system --path-prefix / --dest-name default-petclinic-80 --dest-namespace gloo-system
```

### Local Cluster Adjustments
As we are testing in a local cluster, add the following to your `/etc/hosts` file:
```

> 127.0.0.1 dex.gloo-system.svc.cluster.local

Reviewer: can we make this a command that's actually copy/pasteable?

i.e.,
```suggestion
echo "127.0.0.1 dex.gloo-system.svc.cluster.local" >> /etc/hosts
```

```

> The OIDC flow redirects the browser to a login page hosts by dex. This line in the hosts file will allow this flow to work, with 

Reviewer: something about the wording here is off:
```suggestion
The OIDC flow redirects the browser to a login page hosted by dex. This line in the hosts file will allow this flow to work, with 
```
